File: .agents/teamwork_preview_explorer_quiz_m2_1/parse_lessons.py
import os
import re
import json

project_root = r"C:\Users\UTHtest\.gemini\antigravity\worktrees\network\fix-lesson-completion-logic"
migrations_dir = os.path.join(project_root, "supabase", "migrations")
seed_sql_path = os.path.join(project_root, "src", "app", "data", "seed.sql")

lessons = {}

# Clean SQL quotes
def clean_sql_string(val):
    if not val:
        return ""
    val = val.strip()
    if val.startswith("'") and val.endswith("'"):
        val = val[1:-1]
    # Replace escaped single quotes
    val = val.replace("''", "'")
    return val

# Parse seed.sql
print("Parsing seed.sql...")
try:
    with open(seed_sql_path, "r", encoding="utf-8") as f:
        content = f.read()
    
    # Simple regex to find the INSERT INTO lessons VALUES block
    # We can match everything between INSERT INTO lessons ... VALUES and the next semicolon
    insert_pattern = re.compile(r"INSERT INTO lessons\s*\([^)]+\)\s*VALUES\s*(.*?);", re.DOTALL | re.IGNORECASE)
    match = insert_pattern.search(content)
    if match:
        values_block = match.group(1)
        # Split values block by rows, which are typically formatted like ('val', 'val', ...),
        # We can split by '), ('
        # To be safe, we split by '), \n(' or similar, or use a state machine
        in_row = False
        row_content = ""
        rows = []
        i = 0
        while i < len(values_block):
            char = values_block[i]
            if char == '(' and not in_row:
                in_row = True
                row_content = ""
            elif char == ')' and in_row:
                in_row = False
                rows.append(row_content)
            elif in_row:
                row_content += char
            i += 1
        
        for row in rows:
            # Parse columns, respecting quotes
            cols = []
            curr = ""
            in_quote = False
            i = 0
            while i < len(row):
                c = row[i]
                if c == "'" and i + 1 < len(row) and row[i+1] == "'":
                    curr += "'"
                    i += 2
                    continue
                elif c == "'":
                    in_quote = not in_quote
                    curr += "'"
                elif c == "," and not in_quote:
                    cols.append(curr.strip())
                    curr = ""
                else:
                    curr += c
                i += 1
            cols.append(curr.strip())
            
            if len(cols) >= 7:
                lid = clean_sql_string(cols[0])
                title_th = clean_sql_string(cols[2])
                title_en = clean_sql_string(cols[3])
                content_th = clean_sql_string(cols[4])
                content_en = clean_sql_string(cols[5])
                ltype = clean_sql_string(cols[6])
                lessons[lid] = {
                    "id": lid,
                    "title_th": title_th,
                    "title_en": title_en,
                    "content_th": content_th,
                    "content_en": content_en,
                    "lesson_type": ltype,
                    "source": "seed.sql"
                }
except Exception as e:
    print(f"Error parsing seed.sql: {e}")

# Parse migrations
print("Parsing migrations...")
migration_files = sorted([f for f in os.listdir(migrations_dir) if f.endswith(".sql")])

for file in migration_files:
    file_path = os.path.join(migrations_dir, file)
    with open(file_path, "r", encoding="utf-8") as f:
        sql = f.read()
    
    # Let's split by statement (ignoring semicolons in quotes)
    statements = []
    current = ""
    in_quote = False
    i = 0
    while i < len(sql):
        c = sql[i]
        if c == "'" and i + 1 < len(sql) and sql[i+1] == "'":
            current += "''"
            i += 2
            continue
        elif c == "'":
            in_quote = not in_quote
            current += "'"
        elif c == ";" and not in_quote:
            statements.append(current)
            current = ""
        else:
            current += c
        i += 1
    if current.strip():
        statements.append(current)
        
    for stmt in statements:
        stmt = stmt.strip()
        stmt_upper = stmt.upper()
        if stmt_upper.startswith("INSERT INTO PUBLIC.LESSONS") or stmt_upper.startswith("INSERT INTO LESSONS"):
            # Find VALUES
            val_match = re.search(r"VALUES\s+(.*)$", stmt, re.DOTALL | re.IGNORECASE)
            if val_match:
                vals = val_match.group(1)
                in_row = False
                row_content = ""
                rows = []
                j = 0
                while j < len(vals):
                    char = vals[j]
                    if char == '(' and not in_row:
                        in_row = True
                        row_content = ""
                    elif char == ')' and in_row:
                        in_row = False
                        rows.append(row_content)
                    elif in_row:
                        row_content += char
                    j += 1
                for row in rows:
                    cols = []
                    curr = ""
                    in_quote = False
                    j = 0
                    while j < len(row):
                        c = row[j]
                        if c == "'" and j + 1 < len(row) and row[j+1] == "'":
                            curr += "'"
                            j += 2
                            continue
                        elif c == "'":
                            in_quote = not in_quote
                            curr += "'"
                        elif c == "," and not in_quote:
                            cols.append(curr.strip())
                            curr = ""
                        else:
                            curr += c
                        j += 1
                    cols.append(curr.strip())
                    if len(cols) >= 7:
                        lid = clean_sql_string(cols[0])
                        title_th = clean_sql_string(cols[2])
                        title_en = clean_sql_string(cols[3])
                        content_th = clean_sql_string(cols[4])
                        content_en = clean_sql_string(cols[5])
                        ltype = clean_sql_string(cols[6])
                        lessons[lid] = {
                            "id": lid,
                            "title_th": title_th,
                            "title_en": title_en,
                            "content_th": content_th,
                            "content_en": content_en,
                            "lesson_type": ltype,
                            "source": file
                        }
        elif stmt_upper.startswith("UPDATE PUBLIC.LESSONS") or stmt_upper.startswith("UPDATE LESSONS"):
            # Update statement
            # UPDATE [public.]lessons SET ... WHERE id = '...';
            where_match = re.search(r"WHERE\s+id\s*=\s*'([^']+)'", stmt, re.IGNORECASE)
            if not where_match:
                continue
            lid = where_match.group(1)
            if lid not in lessons:
                lessons[lid] = {
                    "id": lid,
                    "title_th": "",
                    "title_en": "",
                    "content_th": "",
                    "content_en": "",
                    "lesson_type": "",
                    "source": file
                }
            
            # Extract SET assignments
            set_match = re.search(r"SET\s+(.*?)\s+WHERE", stmt, re.DOTALL | re.IGNORECASE)
            if not set_match:
                continue
            set_part = set_match.group(1)
            
            # Parse SET assignments
            assignments = {}
            j = 0
            in_quote = False
            curr_key = ""
            curr_val = ""
            in_val = False
            while j < len(set_part):
                c = set_part[j]
                if c == "'" and j + 1 < len(set_part) and set_part[j+1] == "'":
                    if in_val:
                        curr_val += "''"
                    j += 2
                    continue
                elif c == "'":
                    in_quote = not in_quote
                    if in_val:
                        curr_val += "'"
                    j += 1
                    continue
                elif c == "=" and not in_quote and not in_val:
                    in_val = True
                    j += 1
                    continue
                elif c == "," and not in_quote and in_val:
                    assignments[curr_key.strip()] = curr_val.strip()
                    curr_key = ""
                    curr_val = ""
                    in_val = False
                    j += 1
                    continue
                
                if in_val:
                    curr_val += c
                else:
                    curr_key += c
                j += 1
            if curr_key.strip() and in_val:
                assignments[curr_key.strip()] = curr_val.strip()
                
            for col, val in assignments.items():
                col_name = col.lower().strip()
                val_cleaned = clean_sql_string(val)
                if col_name == "title_en":
                    lessons[lid]["title_en"] = val_cleaned
                elif col_name == "title_th":
                    lessons[lid]["title_th"] = val_cleaned
                elif col_name == "content_en":
                    lessons[lid]["content_en"] = val_cleaned
                elif col_name == "content_th":
                    lessons[lid]["content_th"] = val_cleaned
                elif col_name == "lesson_type":
                    lessons[lid]["lesson_type"] = val_cleaned

print(f"Total unique lessons parsed: {len(lessons)}")
with open(os.path.join(os.path.dirname(__file__), "lessons_extracted.json"), "w", encoding="utf-8") as f:
    json.dump(lessons, f, indent=2, ensure_ascii=False)
print("Saved to lessons_extracted.json")

File: .agents/teamwork_preview_explorer_m2_1/parse_lessons.js
const fs = require('fs');
const path = require('path');

const projectRoot = path.resolve(__dirname, '..', '..');
const migrationsDir = path.join(projectRoot, 'supabase', 'migrations');
const seedSqlPath = path.join(projectRoot, 'src', 'app', 'data', 'seed.sql');

// Map of lessonId -> { id, title_en, title_th, content_en, content_th, lesson_type }
const lessons = {};

// Helper to clean quotes/escapes in SQL strings
function cleanSqlString(str) {
  if (!str) return '';
  str = str.trim();
  if (str.startsWith("'") && str.endsWith("'")) {
    str = str.slice(1, -1);
  }
  // Replace double single quotes '' with single quote '
  str = str.replace(/''/g, "'");
  return str;
}

// Parse seed.sql
console.log('Parsing seed.sql...');
try {
  const seedContent = fs.readFileSync(seedSqlPath, 'utf8');
  // Match INSERT INTO lessons ... VALUES ...;
  // Let's do a regex or simple line-by-line parsing
  const lines = seedContent.split('\n');
  let inLessonsInsert = false;
  let valuesBlock = '';
  
  for (let line of lines) {
    if (line.includes('INSERT INTO lessons') && line.includes('VALUES')) {
      inLessonsInsert = true;
      continue;
    }
    if (inLessonsInsert) {
      if (line.trim().endsWith(';')) {
        valuesBlock += ' ' + line.slice(0, -1);
        inLessonsInsert = false;
        parseInsertValues(valuesBlock);
        valuesBlock = '';
      } else {
        valuesBlock += ' ' + line;
      }
    }
  }
} catch (e) {
  console.error('Error parsing seed.sql:', e);
}

function parseInsertValues(block) {
  // block contains list of rows like ('id', 'course_id', 'title_th', 'title_en', 'content_th', 'content_en', 'lesson_type', ...)
  // Let's parse them carefully, handling commas inside quotes.
  // A simple way is to use a state machine to split by rows
  let i = 0;
  let inRow = false;
  let rowContent = '';
  const rows = [];
  
  while (i < block.length) {
    const char = block[i];
    if (char === '(' && !inRow) {
      inRow = true;
      rowContent = '';
    } else if (char === ')' && inRow) {
      // Check if not inside quote (we simplify by assuming row ends at ) followed by , or end of block)
      // Actually, we can check if it's the end of row
      inRow = false;
      rows.push(rowContent);
    } else if (inRow) {
      rowContent += char;
    }
    i++;
  }

  for (let row of rows) {
    // Split row by commas, respecting quotes
    const cols = parseSqlRowCols(row);
    if (cols.length >= 7) {
      const id = cleanSqlString(cols[0]);
      const title_th = cleanSqlString(cols[2]);
      const title_en = cleanSqlString(cols[3]);
      const content_th = cleanSqlString(cols[4]);
      const content_en = cleanSqlString(cols[5]);
      const lesson_type = cleanSqlString(cols[6]);
      
      lessons[id] = { id, title_th, title_en, content_th, content_en, lesson_type };
    }
  }
}

function parseSqlRowCols(row) {
  const cols = [];
  let currentVal = '';
  let inQuote = false;
  let i = 0;
  
  while (i < row.length) {
    const char = row[i];
    if (char === "'" && row[i+1] === "'") {
      // Escaped single quote
      currentVal += "'";
      i += 2;
      continue;
    }
    if (char === "'") {
      inQuote = !inQuote;
      currentVal += "'";
      i++;
      continue;
    }
    if (char === ',' && !inQuote) {
      cols.push(currentVal.trim());
      currentVal = '';
    } else {
      currentVal += char;
    }
    i++;
  }
  cols.push(currentVal.trim());
  return cols;
}

// Parse migrations
console.log('Parsing migrations...');
const migrationFiles = fs.readdirSync(migrationsDir)
  .filter(f => f.endsWith('.sql'))
  .sort(); // Sort by name (timestamp prefix)

for (let file of migrationFiles) {
  const filePath = path.join(migrationsDir, file);
  const sql = fs.readFileSync(filePath, 'utf8');
  
  // Look for INSERT INTO public.lessons or INSERT INTO lessons
  // Also look for UPDATE lessons or UPDATE public.lessons
  // We can write a simple regex parser for UPDATE
  // Format: UPDATE [public.]lessons SET col1 = val1, col2 = val2, ... WHERE id = 'id';
  // Let's parse line by line or statement by statement.
  
  // Split by semicolon (ignoring semicolons in quotes is tricky but let's do state machine)
  const statements = splitStatements(sql);
  for (let stmt of statements) {
    stmt = stmt.trim();
    if (stmt.toUpperCase().startsWith('INSERT INTO PUBLIC.LESSONS') || stmt.toUpperCase().startsWith('INSERT INTO LESONS') || stmt.toUpperCase().startsWith('INSERT INTO LESSONS')) {
      // Parse insert
      const match = stmt.match(/VALUES\s+([\s\S]+)$/i);
      if (match) {
        parseInsertValues(match[1]);
      }
    } else if (stmt.toUpperCase().startsWith('UPDATE PUBLIC.LESSONS') || stmt.toUpperCase().startsWith('UPDATE LESSONS')) {
      // Parse update
      parseUpdateStatement(stmt, file);
    }
  }
}

function splitStatements(sql) {
  const statements = [];
  let current = '';
  let inQuote = false;
  let i = 0;
  while (i < sql.length) {
    const char = sql[i];
    if (char === "'" && sql[i+1] === "'") {
      current += "''";
      i += 2;
      continue;
    }
    if (char === "'") {
      inQuote = !inQuote;
      current += "'";
      i++;
      continue;
    }
    if (char === ';' && !inQuote) {
      statements.push(current);
      current = '';
    } else {
      current += char;
    }
    i++;
  }
  if (current.trim()) {
    statements.push(current);
  }
  return statements;
}

function parseUpdateStatement(stmt, filename) {
  // UPDATE [public.]lessons SET ... WHERE id = '...';
  // Let's find WHERE id = '...'
  const whereMatch = stmt.match(/WHERE\s+id\s*=\s*'([^']+)'/i);
  if (!whereMatch) return;
  const id = whereMatch[1];
  
  if (!lessons[id]) {
    lessons[id] = { id, title_en: '', title_th: '', content_en: '', content_th: '', lesson_type: '' };
  }
  
  // Extract SET assignments
  // Format: SET col1 = val1, col2 = val2, ...
  const setPartMatch = stmt.match(/SET\s+([\s\S]+?)\s+WHERE/i);
  if (!setPartMatch) return;
  const setPart = setPartMatch[1];
  
  // Parse assignments like col = val
  const assignments = parseSetAssignments(setPart);
  for (let [col, val] of Object.entries(assignments)) {
    const colName = col.toLowerCase().trim();
    if (colName === 'title_en') lessons[id].title_en = cleanSqlString(val);
    if (colName === 'title_th') lessons[id].title_th = cleanSqlString(val);
    if (colName === 'content_en') lessons[id].content_en = cleanSqlString(val);
    if (colName === 'content_th') lessons[id].content_th = cleanSqlString(val);
    if (colName === 'lesson_type') lessons[id].lesson_type = cleanSqlString(val);
  }
}

function parseSetAssignments(setPart) {
  const assignments = {};
  let i = 0;
  let inQuote = false;
  let currentKey = '';
  let currentVal = '';
  let inVal = false;
  
  while (i < setPart.length) {
    const char = setPart[i];
    if (char === "'" && setPart[i+1] === "'") {
      if (inVal) currentVal += "''";
      i += 2;
      continue;
    }
    if (char === "'") {
      inQuote = !inQuote;
      if (inVal) currentVal += "'";
      i++;
      continue;
    }
    
    if (char === '=' && !inQuote && !inVal) {
      inVal = true;
      i++;
      continue;
    }
    
    if (char === ',' && !inQuote && inVal) {
      assignments[currentKey.trim()] = currentVal.trim();
      currentKey = '';
      currentVal = '';
      inVal = false;
      i++;
      continue;
    }
    
    if (inVal) {
      currentVal += char;
    } else {
      currentKey += char;
    }
    i++;
  }
  if (currentKey.trim() && inVal) {
    assignments[currentKey.trim()] = currentVal.trim();
  }
  return assignments;
}

console.log(`Total lessons parsed: ${Object.keys(lessons).length}`);
fs.writeFileSync(
  path.join(__dirname, 'lessons_parsed.json'),
  JSON.stringify(lessons, null, 2)
);
console.log('Saved lessons_parsed.json');
